File: Services/pkg/role_permission/cache.go
package role_permission

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"
	"github.com/gofiber/storage/redis"
)

const (
	cacheTTL     = 15 * time.Minute
	cacheTTLList = 1 * time.Minute
)

type rolePermissionCache struct {
	store *redis.Storage
}

func newRolePermissionCache(store *redis.Storage) *rolePermissionCache {
	return &rolePermissionCache{store: store}
}

func (c *rolePermissionCache) GetByRoleID(roleID uint) ([]models.RolePermission, bool) {
	if c.store == nil {
		return nil, false
	}
	key := fmt.Sprintf("%s:%d", models.PkgRolePermissionGetByRoleID, roleID)

	bytes, err := c.store.Get(key)
	if err != nil || len(bytes) == 0 {
		return nil, false
	}

	var rolePermissions []models.RolePermission
	if err := json.Unmarshal(bytes, &rolePermissions); err != nil {
		return nil, false
	}
	return rolePermissions, true
}

func (c *rolePermissionCache) SetByRoleID(roleID uint, rolePermissions []models.RolePermission) {
	if c.store == nil {
		return
	}
	go func(rps []models.RolePermission) {
		key := fmt.Sprintf("%s:%d", models.PkgRolePermissionGetByRoleID, roleID)
		if bytes, err := json.Marshal(rps); err == nil {
			_ = c.store.Set(key, bytes, utils.RandomJitter(cacheTTLList))
		}
	}(rolePermissions)
}

func (c *rolePermissionCache) InvalidateByRoleID(roleID uint) {
	if c.store == nil {
		return
	}
	go func() {
		key := fmt.Sprintf("%s:%d", models.PkgRolePermissionGetByRoleID, roleID)
		_ = c.store.Delete(key)
	}()
}

File: hono-clean-architecture/pkg/role_permission/cache.ts
import Redis from "ioredis";
import { RolePermission } from "../models";
import { randomJitter } from "../utils";

const CACHE_TTL = 15 * 60 * 1000;

export class RolePermissionCache {
  private store: Redis | null;

  constructor(store: Redis | null) {
    this.store = store;
  }

  async get(id: number): Promise<RolePermission | null> {
    if (!this.store) return null;

    const key = `pkg:rolepermission:get:${id}`;
    const data = await this.store.get(key);

    if (!data) return null;

    try {
      return JSON.parse(data) as RolePermission;
    } catch {
      return null;
    }
  }

  async set(rolePermission: RolePermission): Promise<void> {
    if (!this.store || !rolePermission) return;

    const key = `pkg:rolepermission:get:${rolePermission.id}`;
    const ttl = randomJitter(CACHE_TTL);

    await this.store.set(key, JSON.stringify(rolePermission), "PX", ttl);
  }

  async invalidate(id: number): Promise<void> {
    if (!this.store) return;
    await this.store.del(`pkg:rolepermission:get:${id}`);
  }

  async invalidateAllLists(): Promise<void> {
    if (!this.store) return;

    const pattern = `pkg:rolepermission:list:*`;
    const keys = await this.store.keys(pattern);

    if (keys.length > 0) {
      await this.store.del(...keys);
    }
  }
}

export function newRolePermissionCache(store: Redis | null): RolePermissionCache {
  return new RolePermissionCache(store);
}
